### ISSUE TITLE
support DNG format input (including bayer pattern)

### ISSUE THREAD

**Issue by StephanBusch** [FLIF-hub/FLIF#111]
please add support for DNG as input format (lossless JPEG)
including mosaiced raw (bayer pattern array)
and during FLIF compression leave image mosaiced which means to not convert to RGB.

Explanations:

According to Dave Coffin, author of DCRAW, such a lossless compression could be done this way:

„Run "dcraw -i orig.raw" and printf() the "data_offset" value. This is where the raw pixels begin.
Figure out where the raw pixels end. This requires a little extra logic in dcraw.c, which is not written yet.
Copy orig.raw to meta.raw, and zero out the raw pixel data.
Do "dcraw -E -4 -c orig.raw | tail -n +3 | dd conv=swab > pixels.raw" IOW, dump raw pixels, cut off the header, and byte-swap.
Compress meta.raw and pixels.raw using appropriate algorithms, and save them in a single wrapper file. The zeroed-out raw pixels in meta.raw will collapse to almost nothing.

To decode these images:

Extract meta.raw and pixels.raw from the wrapper and decompress.
Run "dcraw -I pixels.raw meta.raw". Dcraw reads the raw pixels from pixels.raw and everything else from meta.raw.

Recreating the original raw file would be a little more complicated, since you'd have to reverse each "load_raw" function into a "save_raw" function. Or just wrap orig.raw instead of meta.raw,
(Adobe DNG converter allows this) but that won't save any disk space.“

Klaus Post, author of RawSpeed, has written such a splitter and compressor based on his program RawSpeed.. It is available here (including sources). Tests showed that using Delta and ZigZag-Encoding before using general-purpose compression algorithm would improve compression.
According to Matthias Stirner, author of packJPG/packMP3/packPNM, the Hilbert-Scan „leads to better results in every case tested“.

Thomas Richter wrote that „The principles you need are "decorrelation" and "energy compaction", with a suitable statistical model. JPEG-LS is a nice starter. Try to combine that with a better cross-component correlation, make use of the position of the samples of the Bayer patterns to improve the prediction, probably try a non-local matching of patterns. You probably also want to read the Weinberger paper on JPEG LS which describes a lot of the techniques, and in specific one important property of natural data prediction residuals: These are almost always two-sided Laplacian (or close to that), a family of probability distributions "stable under convolution". The reason why LS works so nice is because it makes use of this property. In the end, you'll likely notice that with a pretty primitive prediction, you get about 1:2 compression. With something very sophisticated, you get a bit above 1:2. (-; Reason is that most data you got there is noise, and not actual image data, and noise cannot be compressed very well. It's Laplacian noise, though.“

**Comment by psykauze** [comment]
I don't think that Flif will claim a bit-exact file compression but only a pixel-exact image compression.

Also, if jonsneyers wants a full meta-data support to his file-format, I think he does not care if the Author Tag is at the same position than the original file. Or if the PNG file has the same file size than before.

Concerning the DNG input support without debayer, flif support it yet (with the raw2flif script). For now, it just using a dcraw command line but there's not yet a solution to convert it back to Raw.

**Comment by jonsneyers** [comment]
Thanks for this information!
The `raw2flif` script already takes care of some of this, but I guess it does not put enough metadata in the .flif archive to allow full reconstruction. I'll have to take a look at the Klaus Post splitter. The link is missing though...

**Comment by StephanBusch** [comment]
I wouldn't care for a bi-exact file compression as long as the FLIF files are pixel-exact compared to original RAW. All we need now are Raw Converters such as DCRAW, Photoshop, DxO, Capture One
that directly read the FLIF format for further processing/de-mosaicing.
If FLIF would read DNG format, conversions could be directly done from within FLIF
and people would get used to developing good-compressed FLIF rather than weak-compressed DNG..

the binaries of Klaus Post RawSplitter are here:
http://download.klauspost.com/rawcompress/

the sources are also at GitHub:
https://github.com/klauspost/rawspeed/tree/RawspeedCompress

Does the RGGB file contain the original (mosaiced) pixel data of a raw?

**Comment by psykauze** [comment]
The RGGB file format is extracted as an uncompressed greyscale without any gamma correction nor unmosaicing from the raw file via dcraw (DNG, NEF, etc...)
For now flif is assuming the matrix pattern is RG1G2B.
There is not yet a FLIF to Raw functionality, only a Flif to mosaiced greyscale.

**Comment by StephanBusch** [comment]
I made a compression test for Camera Raw images.

Every FLIF result consists of FLIF size + JPEG thumbnails that are embedded in original RAW.
For Sigma Raw, DCRAW -E -4 created a PPM file because Sigma does not use Bayer pattern.
http://www.squeezechart.com/camera-raw.html

When compressing those RGGB, the bpp value is wrong each time.

**Comment by jonsneyers** [comment]
If you are using `raw2flif`, then at least some of the metadata and embedded thumbnails/previews should end up in the resulting flif file (it uses `exiv2` to extract that stuff). Anyway the idea of progressive flif is that you don't really need previews, you just decode a small prefix (you could consider the RGGB values as belonging to the same 'big' pixel if you only want a preview)...

The bpp values it gives are just the 16 bits of the PGM dcraw produces. To deduce the actual bpp per color channel you could look at the per-channel bounds (if you add enough `-vvvv`, the BND intervals are shown).

**Comment by StephanBusch** [comment]
can anyone please instruct me how to use raw2flif on a windows machine so that I can redo the tests?
And is there a way to save further space by improving compression on RGGB and/or taking care of some raw that only contain 8, 10, 12 or 14 bit which are currently stored in a 16 bit PGM?
I would also vote for no thumbnails.

**Comment by jonsneyers** [comment]
`raw2flif` essentially simply renames the output of `dcraw -E -4` from .pgm to .rggb to ensure that FLIF knows how to interpret the input data. It flifs the raw rggb data. Then it adds the metadata sidecar files extracted using `exiv2` and rolls all of that into an ar.

FLIF always takes the bounds of the input channels into account, so it works with the effective bit depths, not the declared ones.
